core/lib/types/src/storage/writes/mod.rs

```diff
@@ -0,0 +1,499 @@
+pub struct StateDiffRecord {
+    /// address state diff occured at
+    pub address: Address,
+    /// storage slot key updated
+    pub key: U256,
+    /// derived_key == Blake2s(bytes32(address), key)
+    pub derived_key: [u8; 32],
+    /// index in tree of state diff
+    pub enumeration_index: u64,
+    /// previous value
+    pub initial_value: U256,
+    /// updated value
+    pub final_value: U256,
+}
+
+impl StateDiffRecord {
+    // Serialize into byte representation.
+    fn encode(&self) -> [u8; STATE_DIFF_RECORD_SIZE] {
+        let mut encoding = [0u8; STATE_DIFF_RECORD_SIZE];
+        let mut offset = 0;
+        let mut end = 0;
+
+        end += 20;
+        encoding[offset..end].copy_from_slice(self.address.as_fixed_bytes());
+        offset = end;
+
+        end += 32;
+        self.key.to_big_endian(&mut encoding[offset..end]);
+        offset = end;
+
+        end += 32;
+        encoding[offset..end].copy_from_slice(&self.derived_key);
+        offset = end;
+
+        end += 8;
+        encoding[offset..end].copy_from_slice(&self.enumeration_index.to_be_bytes());
+        offset = end;
+
+        end += 32;
+        self.initial_value.to_big_endian(&mut encoding[offset..end]);
+        offset = end;
+
+        end += 32;
+        self.final_value.to_big_endian(&mut encoding[offset..end]);
+        offset = end;
+
+        debug_assert_eq!(offset, encoding.len());
+
+        encoding
+    }
+
+    pub fn encode_padded(&self) -> [u8; PADDED_ENCODED_STORAGE_DIFF_LEN_BYTES] {
+        let mut extended_state_diff_encoding = [0u8; PADDED_ENCODED_STORAGE_DIFF_LEN_BYTES];
+        let packed_encoding = self.encode();
+        extended_state_diff_encoding[0..packed_encoding.len()].copy_from_slice(&packed_encoding);
+
+        extended_state_diff_encoding
+    }
+
+    /// Decode bytes into StateDiffRecord
+    pub fn from_slice(data: &[u8]) -> Self {
+        assert_eq!(data.len(), 156);
+
+        Self {
+            address: Address::from_slice(&data[0..20]),
+            key: U256::from(&data[20..52]),
+            derived_key: data[52..84].try_into().unwrap(),
+            enumeration_index: u64::from_be_bytes(data[84..92].try_into().unwrap()),
+            initial_value: U256::from(&data[92..124]),
+            final_value: U256::from(&data[124..156]),
+        }
+    }
+
+    /// compression follows the following algo:
+    /// 1. if repeated write:
+    ///      entry <- enumeration_index || value
+    /// 2. if initial write:
+    ///      entry <- blake2(bytes32(address), key) || value
+    /// size:
+    ///      initial:  64 bytes
+    ///      repeated: 40 bytes
+    ///      before:  156 bytes for each
+    pub fn compress(&self) -> Vec<u8> {
+        let mut comp_state_diff = match self.enumeration_index {
+            0 => self.derived_key.to_vec(),
+            enumeration_index if enumeration_index < u32::MAX.into() => {
+                (self.enumeration_index as u32).to_be_bytes().to_vec()
```

Comment:
I wonder if we should check that this matches BYTES_PER_ENUMERATION_INDEX (some debug_assert?)